### PR TITLE
Relative correctness policy in Caboose

### PR DESCRIPTION
**Pull request by aarshkshah1992**
We don't want to evict nodes if their correctness is similar to other nodes even if their absolute correctness is below a certain random threshold as software bugs can cause absolute correctness to dip which will result in us evicting legit nodes.

### CODE REVIEW

File: tieredhashing/tiered_hashing.go [hunk 1]
```diff
@@ -418,10 +423,23 @@ func (t *TieredHashing) isCorrectnessPolicyEligible(perf *NodePerf) (float64, bo
 		return result
 	})
 
+	averageSuccess := t.OverAllCorrectnessDigest.Reduce(func(w rolling.Window) float64 {
```
Comment (willscott) [inline tieredhashing/tiered_hashing.go:426]: is there any way to memo-ize this value / this function?

doing the reduction over 1000 data points on every single request when the general 'overall correctness' ought to be fairly stable  is a lot of duplicated work.

we can try in staging but i am worried this will be a cpu bottleneck when run at scale as it's written

File: tieredhashing/tiered_hashing.go [hunk 2]
```diff
@@ -421,7 +443,7 @@ func (t *TieredHashing) isCorrectnessPolicyEligible(perf *NodePerf) (float64, bo
 	// should satisfy a certain minimum percentage
 	pct := totalSuccess / perf.NCorrectnessDigest * 100
 
-	return pct, pct >= t.cfg.CorrectnessPct
+	return pct, (t.AverageCorrectnessPct - pct) < t.cfg.CorrectnessThreshold
```
Comment (willscott) [inline tieredhashing/tiered_hashing.go:446]: is this backwards? ( lets put in a comment of what the boolean means)

if average correctness is 80%
this node correctness is 20%

0.8 - 0.2 = 0.6
0.6 is more than the 20% threshold.

Comment (aarshkshah1992) [reply tieredhashing/tiered_hashing.go:446]: Faceplam. You are right.

Comment (aarshkshah1992) [reply tieredhashing/tiered_hashing.go:446]: @willscott It is correct actually. This boolean is true when we do NOT want to remove a node from the pool i.e. true means node is eligible for the correctness policy.

Comment (aarshkshah1992) [reply tieredhashing/tiered_hashing.go:446]: In your example, bool will be false and so node will be kicked out.

File: tieredhashing/tiered_hashing.go [hunk 3]
```diff
@@ -27,7 +27,7 @@ const (
 	reasonCorrectness = "correctness"
 
 	// use rolling windows for latency and correctness calculations
-	latencyWindowSize     = 100
+	latencyWindowSize     = 50
```
Comment (willscott) [inline tieredhashing/tiered_hashing.go:30]: this seems unrelated to the larger correctness change? can we do this as a different PR / test them independently?

File: tieredhashing/tiered_hashing.go [hunk 4]
```diff
@@ -31,12 +31,16 @@ const (
...
+	correctnessThreshold = float64(25)
 
 	// helps shield nodes against bursty failures
 	failureDebounce = 2 * time.Second
-	removalDuration = 3 * time.Hour
+	removalDuration = 6 * time.Hour
+)
+
+var (
+	goLogger = golog.Logger("caboose-pool")
```
Comment (willscott) [inline tieredhashing/tiered_hashing.go:43]: caboose-tiered-hashing ?

File: tieredhashing/tiered_hashing.go (hunk 2 above)
Comment (willscott) [reply tieredhashing/tiered_hashing.go:457]: can you put in the comment explaining this?

**Review by willscott — APPROVED** [review]

**Comment by aarshkshah1992** [comment]
@willscott Whenever you are ready.

**Comment by aarshkshah1992** [comment]
@willscott Addressed your last couple of comments.
